Working Band Aid Fix for time

leavesandlemmas opened the pull request:
Okay, so this pull request effectively removes `time` as a variable. All of the crop model that use `time` assume that it is equal to `time = day_of_year + hour_of_day / 24`. In fact, the `solar_position_michalsky` module recomputes the day of the year and hour from the provided `time` variable. So it now takes `hour` and `doy` as inputs, rather than `time`.  

Calling `add_time_to_weather_data` in `run_biocro` effectively made it a requirement to specify `time`, or the `doy` and `hour` variables, even when the model doesn't require it. All instances of `time` in the crop models could be replaced with `doy` without any loss of generality.  So that's what I did.

Most of the other changes are required to get the tests to work again. 

Edit by Justin: This is a continuation of #135.

leavesandlemmas commented:
Okay, apparently not working because of the vignettes.

justinmcgrath commented:
It's already know that these values are redundant, and in the context of the crop growth models it doesn't very much matter which way the calculations go. This does not appear to address the possibility of mismatches between the times and the timestep or using the same time point in multiple rows of the driver. What is the purpose of doing this?

leavesandlemmas commented:
> It's already know that these values are redundant, and in the context of the crop growth models it doesn't very much matter which way the calculations go. This does not appear to address the possibility of mismatches between the times and the timestep or using the same time point in multiple rows of the driver. What is the purpose of doing this?

Well, it at least removes the hard coded assumption that `time` is `day_of_year + hour_of_day/24`.  Now, time can have any units I suppose.

leavesandlemmas commented:
Okay, I added back the time variable check. Now, any user must provide drivers with a time variable, but it must be spaced by timestep.  The other modules were rewritten to not use time explicitly. I modified the `add_time_to_weather_data` so that it adds a `time` variable equal to the number of hours elapsed since midnight Jan 1 of the year.

justinmcgrath commented:
Here's a summary of a conversation with @leavesandlemmas.

The other PR is no longer needed and can be closed.

In order to ensure that the drivers and solver are consistent, the time in the drivers when using crop model needs to be in hours, so that it can match the time step.

In order order to do that, a number of modules need to be changed to match units.

Overall, it sounds like it's the right track, but it would be good to determine whether we need to change time to hours for crop models, or if there isn't some other option.

He removed the the code to add time in run_biocro, with the expectation that users would call this themselves when using crop models. I don't know how I feel about that. It is more in line with run_biocro being general purpose, but 95% of what we use run_biocro for is crop models, and the convenience is nice and because of the check for time, it doesn't make it any less general. The only cost is a cheap if check and some extra code.

gsrohde commented:
I really don't see much utility in changing `time` to be in units of hours rather than days except that possibly it makes `check_time_is_correct_format` slightly easier to write.  If the main concern is the units disparity between `time`, which up to now has had units of days, and `timestep`, which for all differential modules of significant interest for running plant growth simulations (namely, all differential modules other than `harmonic_oscillator`) must be in units of hours, then it would be much simpler simply to rename the `time` parameter to something like `fractional_doy` (or, if that seems too ambiguous, maybe the somewhat cumbersome `doy_including_fractional_portion`).

leavesandlemmas commented:
My understanding is that @justinmcgrath would like BioCro to not assume any units for `timestep` while also checking that the drivers are time-series spaced by `timestep`. But to check the drivers are correctly spaced, BioCro has to assume something about how the quantities `timestep` and `time` are related.  The simplest assumption is that the `time` variable always has the same units as `timestep` so that the `time` variable should jump by `timestep`. The  `harmonic_oscillator` module tests assume this is the case (although done in a roundabout way of setting `time = hour`). 

I do think renaming `time` to `fractional_doy` is a good idea as `fractional_doy` hints at the correct units. Originally, I just eliminated the `time` variable so it wasn't possible to confuse its meaning. But without some sort of `time` variable its hard to check whether the drivers are sequential in time.

gsrohde reviewed (commented):
As I've already stated elsewhere, I question the utility of changing `time` to be in units of hours rather than days.  It seems like a lot of work for very little, if any benefit.  Nevertheless, I've made a few comments upon the code in case others strongly feel that this change _should_ be made.

leavesandlemmas commented on R/add_time_to_weather_data.R:
```diff
@@ -4,8 +4,10 @@ add_time_to_weather_data <- function(drivers)
+        day_fraction <- drivers$hour
+        # time = 0 is equivalent to Jan 1, 00:00:00
+        days_per_hour <- 24
+        drivers['time'] <- (drivers$doy-1)*days_per_hour  + day_fraction
```
I don't think I am particularly consistent with the use of `$` vs. brackets. For whatever reason (aesthetics I guess), I tend to prefer brackets over `$` but often use `$` out of habit. 

I wonder if the partial matching of `$` also applies to assignment...  I don't really understand the reason for the multiple indexing operators (other than maybe `$` allows slightly less typing).  A single bracket returns a slice while a double bracket returns an entry, right?  So if `df` is a dataframe then `df['x']` is a single column dataframe while `df[['x']]` is a vector ? 

It probably doesn't matter for assignment since assigning a vector to `df['x']` just promotes the vector into a dataframe or something. Not sure though, my knowledge of R's inner workings is a bit fuzzy.

gsrohde replied:
> I don't think I am particularly consistent with the use of `$` vs. brackets. For whatever reason (aesthetics I guess), I tend to prefer brackets over `$` but often use `$` out of habit.

I don't really have a strong opinion about using one versus the other.  But the fact that you changed it called it to my attention, and I wondered about it, and then I went on to wonder why there are so many uses of "$" in our R code, seemingly contradicting the guideline I cited above.  (A quick grep in the R directory turns up 45 lines containing "[[" (`grep "\[\[" *.R | wc -l`) versus 21 containing "$" (`grep "\\\$" *.R | wc -l`).)  My guess is that the guideline (like a New Year's resolution) is meant to apply to new code and that it wasn't thought worthwhile changing already-written code.  I'm not even all that concerned, though, that we aren't consistent (“A foolish consistency is the hobgoblin of little minds ..." —Ralph Waldo Emerson), but if one is going to use the `$` operator and the bracket operator interchangeably, it would be clearer to use the bracket operator that is `$`'s exact equivalent (mod partial completion)—namely, double brackets (`[[`).
 
> I wonder if the partial matching of `$` also applies to assignment

From the help page for `$` (obtained using ``?`$` ``): "When an index expression appears on the left side of an assignment (known as subassignment) then that part of x [the sample usage example is `x$name <- value`] is set to the value of the right hand side of the assignment. In this case no partial matching of character indices is done, and the left-hand-side is coerced as needed to accept the values."

> I don't really understand the reason for the multiple indexing operators (other than maybe `$` allows slightly less typing). 

Likely.  Or perhaps `$` predates `[[`, and at some point it was realized that partial matching was problematic, and `[[` was introduced as a way of being more precise without changing the behavior of existing code.

> A single bracket returns a slice while a double bracket returns an entry, right? 

Sort of sounds right, though I'm not clear exactly what you mean by "slice" and "entry".  (The R manuals barely use the term "slice": It doesn't appear at all in [An Introduction to R](https://cran.r-project.org/doc/manuals/r-release/R-intro.html) or in [R Internals](https://cran.r-project.org/doc/manuals/r-release/R-ints.html).  It appears in the [R Language Definition](https://cran.r-project.org/doc/manuals/r-release/R-lang.html), but only once, and in an informal way, when it speaks of taking a two-dimensional slice of a three-dimensional matrix using a construct such as `m[2, , ]`.  [The R Reference Index](https://cran.r-project.org/doc/manuals/r-release/fullrefman.pdf), a compendium of all the help pages, uses "slice" several times, mostly to mean other things (such as slices in a pie graph or in connection with the `slice.index` function).  When it is used to mean something akin to what you mean, it is in the help for the _Extract_ operators (precisely the operators we are talking about), but even there, it is used only in connection with using logical vectors or negative integers to extract portions of an object.  Again, the usage is informal and never precisely defined, leaving the reader to guess exactly what is meant.

The [Tidyverse](https://www.tidyverse.org/) package _dplyr_ contains a function `slice` that works specifically on dataframes and dataframe-like objects to extract rows (horizontal slices).  A different function `select` is used to extract columns.

(_Tidyverse_, as best as I can tell, is an extensive system of packages written by Hadley Wickham et al. to smooth out some of the quirks in R syntax, allowing the user to interact with R objects in a way that is more intuitive and natural (and convenient), an admirable goal.  My impression, though, is that whole system is somewhat bloated and requires significant investment in learning what almost amounts to a new language.  But perhaps this is unfair (or perhaps my impressions of it are unduly (and unconsciously) colored by Hadley's sharing a surname with the rather dastardly Mr. Wickham in _Pride and Prejuduce_).

A brief but valuable-seeming (I've only skimmed it) discussion of the pros and cons of Tidyverse can be found here: https://bitsofanalytics.org/posts/base-vs-tidy/)

> So if `df` is a dataframe then `df['x']` is a single column dataframe while `df[['x']]` is a vector ?

Yes, precisely.  I often use the `str` operator (which, confusingly, abbreviates "structure", not "string" as it does in many (most?) other languages) to figure out exactly what sort of object I am dealing with.  `class()` is also useful.
> 
> It probably doesn't matter for assignment since assigning a vector to `df['x']` just promotes the vector into a dataframe or something. 

According to the help page, in _subassignment_, as this is called, "part of x [or `drivers`, in this case] is set to the value of the right hand side of the assignment" and "the left-hand-side is coerced as needed to accept the values".

> Not sure though, my knowledge of R's inner workings is a bit fuzzy.

Understandable—R is a very untidy language (as implied by the name _Tidyverse_), and moreover, much of the documentation, at least in the help pages, is at times ungrammatical and poorly written.  The manual pages, I think, are somewhat better.

gsrohde commented:
> My understanding is that @justinmcgrath would like BioCro to not assume any units for `timestep` while also checking that the drivers are time-series spaced by `timestep`. 

Getting rid of this assumption will require rewriting all of biological differential modules because they all implicitly assume that `timestep` is in hours.  And no matter what units we use for  `timestep`, all of the differential modules used together in a simulation will have to assume the same units for timestep.  The `harmonic_oscillator` module is unique in assuming that `timestep` has units of seconds, but this is of little consequence because it is unlikely to ever be used with any of the biological modules.

> But to check the drivers are correctly spaced, BioCro has to assume something about how the quantities `timestep` and `time` are related. The simplest assumption is that the `time` variable always has the same units as `timestep` so that the `time` variable should jump by `timestep`. The `harmonic_oscillator` module tests assume this is the case (although done in a roundabout way of setting `time = hour`).

It seems to me that the sort of drivers used with biological simulations will invariably have some sort of time-stamp associated with each row.  Adding some sort of linearly ordered monotonically increasing value to the driver columns (e.g. "time") does perhaps make it slightly easier to check that successive rows are separated by an amount of time corresponding to `timestep`, but it is hardly necessary.  It is easy, for example, write code to check that a row with `doy = 100` and `hour = 0` presents weather data for an instant in time that is exactly one hour later that presented by a row with `doy = 99` and `hour = 23`.
 
> I do think renaming `time` to `fractional_doy` is a good idea as `fractional_doy` hints at the correct units.

I'm not so sure this is true.  I was assuming it would have units of days, but I have the feeling you are assuming it to have units of hours.

> Originally, I just eliminated the `time` variable so it wasn't possible to confuse its meaning. But without some sort of `time` variable its hard to check whether the drivers are sequential in time.

Not that hard.  All we really need is some unambiguous time specification, and `doy` together with `hour` will suffice (as long as we don't cross years—see my comment at https://github.com/orgs/biocro/discussions/25).

leavesandlemmas commented on R/get_growing_season_climate.r:
```diff
@@ -2,7 +2,9 @@ get_growing_season_climate <- function(climate, threshold_temperature = 0) {
         guess_growing_season_start_and_end_doy(climate, threshold_temperature)
 
-    return(climate[with(climate, doy >= doy_range[1] & doy <= doy_range[2]), ])
+    out <- climate[with(climate, doy >= doy_range[1] & doy <= doy_range[2]), ]
```
I am okay with changing the name of this function, but I am going to defer to Ed or Justin's judgement as changing it could break existing scripts.

eloch216 replied:
This function also predates my involvement with BioCro. I agree that it's a weird name and it would be nice to change it, although that doesn't need to happen in this PR.

I think a "polite" way to rename it would be to leave `get_growing_season_climate` in the package namespace, but change its definition to something like

```R
get_growing_season_climate <- function(...) {
  stop('`get_growing_season_climate` has been renamed to `FFF` for clarity. Please use that function instead.')
}
```

where `FFF` is the new name. Eventually, `get_growing_season_climate` can be removed from the namespace. I have done this a few times with `PhotoGEA` functions and I think it works well. Scripts will break, but they will also explain how to fix themselves. See this file for examples from `PhotoGEA`: https://github.com/eloch216/PhotoGEA/blob/main/R/deprecated.R

leavesandlemmas replied:
My suggestion for a new name: `get_growing_season_weather` or even just `get_growing_season` where the first argument is renamed to `weather` or `weather_data`.

gsrohde replied:
> My suggestion for a new name: `get_growing_season_weather` or even just `get_growing_season` where the first argument is renamed to `weather` or `weather_data`.

My two cents?  `get_growing_season_weather` (although slightly longer than `get_growing_season`) clearly states what it does (no danger that you think you're going to get back a date range rather than the weather data for that date range).

gsrohde replied:
> This function also predates my involvement with BioCro. I agree that it's a weird name and it would be nice to change it, although that doesn't need to happen in this PR.
> 
> I think a "polite" way to rename it would be to leave `get_growing_season_climate` in the package namespace, but change its definition to something like
> 
> ```r
> get_growing_season_climate <- function(...) {
>   stop('`get_growing_season_climate` has been renamed to `FFF` for clarity. Please use that function instead.')
> }
> ```
> 
> where `FFF` is the new name. Eventually, `get_growing_season_climate` can be removed from the namespace. I have done this a few times with `PhotoGEA` functions and I think it works well. Scripts will break, but they will also explain how to fix themselves. See this file for examples from `PhotoGEA`: https://github.com/eloch216/PhotoGEA/blob/main/R/deprecated.R

I just want to note (for future reference) that although you don't the term "deprecated" in your proposed redefinition, you do use it in the PhotoGEA functions.  My understanding has always been that deprecated features continue to work essentially as before except that their use may trigger a warning that the feature is slated to be removed (though sometimes this warning only appears in release notes).  But in the examples you present, you've essentially already removed the functions even though technically you haven't—you've just redefined them.  It's sort of like when you call a number and you get a recording that tells you that the number is no longer in service, though technically, it _is_ in service: it's serving to tell you that it can't be used to connect to the person you intended to reach.

justinmcgrath commented:
> My understanding is that @justinmcgrath would like BioCro to not assume any units for `timestep` while also checking that the drivers are time-series spaced by `timestep`. But to check the drivers are correctly spaced, BioCro has to assume something about how the quantities `timestep` and `time` are related.

A solver shouldn't make assumptions about the units in the model, but the values provided to it must match what the model uses, which covers all of the times in the solver, start, end and step.

>  must be in units of hours, then it would be much simpler simply to rename the time parameter to something like fractional_doy

I'm not sure I follow here. Are you saying do not have a variable called time? I think we want to use that name. But maybe going off of this idea, maybe express the timestep in fractions of a day (so 1/24 to keep using an hour) rather than hours. That would make it so that these ones that expect time as the doy don't need to be changed, but I think it would require changing a lot of other places.

leavesandlemmas commented:
> I'm not sure I follow here. Are you saying do not have a variable called time? I think we want to use that name. But maybe going off of this idea, maybe express the timestep in fractions of a day (so 1/24 to keep using an hour) rather than hours. That would make it so that these ones that expect time as the doy don't need to be changed, but I think it would require changing a lot of other places.

I don't think it works to change time step to `1/24`. The differential modules produces quantities with units in hours. The time step has to have the same units as the differential modules, in which case, we'd have to change their parameters to be in days rather than hours.

I think we should be careful with using a variable called `time` that's the same as the calendar time (or `fractional_doy`) because it creates implicit requirements for the `time` variable that are awkward for the dynamical systems that don't have drivers tied to calendar time (like the `harmonic_oscillator` or the reworked nud model). Given that BioCro is intended for crop models, it's not necessarily bad to assume that `time` is a calendar time, but it should probably be enforced correctly.

eloch216 commented:
My overall thought here is that this PR is implementing several changes at once and might be better as multiple PRs. I know I have been guilty of making very complicated PRs many times in the past. But a new year is a good opportunity to change for the better!

I think a first step should be simply checking that the `time` column of the drivers is (i) ordered and (ii) evenly-spaced. I believe this change should not cause any tests or examples to fail, and implementing it in a dedicated PR would make it easy to verify that this is the case. Along with the check, a new test could be added to make sure that "improper" drivers cause an error to occur.

Changing `time` to a different name or different units is more complicated, as we can see here. It would be easier to understand these changes if they were made after the check described above.

justinmcgrath commented:
Some of this has changed since I was heavily involved with the code, but time, as is, appears to be given as days already. I can see changing it to hours so that it matches the timestep, but it should not be removed. We think of these as functions of time, and the clearest expression of that is a variable named time.

I think there is some miscommunication happening, because unless somewhere in the parts that link solver and model expect specific units, then it doesn't matter whether time is in days or any other units. It should only matter that the set of modules and the solver use the same units. There is some problem with that now, and that is part of what we're trying to work out, but it seems tricky getting there.

In light of that, I agree with taking this in a few steps, starting with what is most obvious, committing that, then moving to the next part.

eloch216 commented:
With PRs https://github.com/biocro/biocro/pull/140, https://github.com/biocro/biocro/pull/143, and https://github.com/biocro/biocro/pull/144, most of the goals of this PR have now been accomplished.

I believe the remaining goal is to change a few modules to use `doy`, `hour`, or `fractional_doy` as inputs instead of `time`. I think this would be useful to do, since these modules more directly depend on `doy`, `hour`, or `fractional_doy` than on `time`. As an added bonus, these modules would be insensitive to any future changes to the meaning of `time`. (For instance, if we fully switch to coherent SI units, `time` would need to be in seconds.) @leavesandlemmas do you want to work on this? I think the solar position module probably wants to use `doy` and `hour`, while the phenology modules want to use `fractional_doy` (and `sowing_fractional_doy`).

leavesandlemmas commented:
> > > With PRs #140, #143, and #144, most of the goals of this PR have now been accomplished.
> > > I believe the remaining goal is to change a few modules to use `doy`, `hour`, or `fractional_doy` as inputs instead of `time`. I think this would be useful to do, since these modules more directly depend on `doy`, `hour`, or `fractional_doy` than on `time`. As an added bonus, these modules would be insensitive to any future changes to the meaning of `time`. (For instance, if we fully switch to coherent SI units, `time` would need to be in seconds.) @leavesandlemmas do you want to work on this? I think the solar position module probably wants to use `doy` and `hour`, while the phenology modules want to use `fractional_doy` (and `sowing_fractional_doy`).
> > 
> > 
> > Sure, I can do it when I've got a sec.
> 
> @leavesandlemmas I can work on this

@eloch216  Go ahead. I am busy with nitrogenAssim at the moment, plus other commitments.

leavesandlemmas commented:
Addressed by other pull requests, like #154  #142 #143

eloch216 commented:
This PR is closed, so I will delete the corresponding branch.
